From 052bfbefd6bcbb37a616b1cee5e00fedef5d7d61 Mon Sep 17 00:00:00 2001
From: Super User <root@localhost.localdomain>
Date: Sat, 16 May 2026 10:15:49 +0800
Subject: [PATCH] vfio/pci: Add pre_load to limit cmask validation

Hardware read-only bits should not be validated.
Fix: Limit cmask to only check emulated bits.

Signed-off-by: QEMU VFIO Team <vfio@qemu.org>
---
 hw/vfio/pci.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/hw/vfio/pci.c b/hw/vfio/pci.c
index b2a07f6bb4..df2d5d775d 100644
--- a/hw/vfio/pci.c
+++ b/hw/vfio/pci.c
@@ -2859,10 +2859,25 @@ static const VMStateDescription vmstate_vfio_display = {
     }
 };
 
+static int vfio_pci_config_pre_load(void *opaque)
+{
+    VFIOPCIDevice *vdev = opaque;
+    PCIDevice *pdev = PCI_DEVICE(vdev);
+    int size = MIN(pci_config_size(pdev), vdev->config_size);
+    int i;
+
+    for (i = 0; i < size; i++) {
+        pdev->cmask[i] &= vdev->emulated_config_bits[i];
+    }
+
+    return 0;
+}
+
 static const VMStateDescription vmstate_vfio_pci_config = {
     .name = "VFIOPCIDevice",
     .version_id = 1,
     .minimum_version_id = 1,
+    .pre_load = vfio_pci_config_pre_load,
     .fields = (const VMStateField[]) {
         VMSTATE_PCI_DEVICE(parent_obj, VFIOPCIDevice),
         VMSTATE_MSIX_TEST(parent_obj, VFIOPCIDevice, vfio_msix_present),
-- 
2.43.0

